zk/stages/stage_sequence_execute.go

```diff
@@ -475,31 +509,70 @@ func SpawnSequencingStage(
 		for _, tx := range addedTransactions {
 			log.Debug(fmt.Sprintf("[%s] Finish block %d with %s transaction", logPrefix, thisBlockNumber, tx.Hash().Hex()))
 		}
+
 		log.Info(fmt.Sprintf("[%s] Finish block %d with %d transactions...", logPrefix, thisBlockNumber, len(addedTransactions)))
+
+		if !cfg.zk.HasExecutors() {
+			// save counters midbatch
+			// here they shouldn't add more to counters other than what they already have
+			// because it would be later added twice
+			counters := batchCounters.CombineCollectorsNoChanges(intermediateUsedCounters, l1InfoIndex != 0)
+
+			err = sdb.hermezDb.WriteBatchCounters(thisBatch, counters.UsedAsMap())
+			if err != nil {
+				return err
+			}
+
+			if err = tx.Commit(); err != nil {
+				return err
+			}
+			tx, err = cfg.db.BeginRw(ctx)
+			if err != nil {
+				return err
+			}
+			// TODO: This creates stacked up deferrals
+			defer tx.Rollback()
+			sdb.SetNewTx(tx)
```

Comment:
would this be better positioned after the write to the datastream?  Just thinking aloud, if the write to the stream fails we will have already committed the block details and counters to the DB.  If we do it after the write to the stream then everything gets commited as one, any issue committing the DB stuff will cause an unwind in the stream anyway which we're already handling.